cpp/src/arrow/util/bit_util.h

```diff
@@ -282,15 +282,19 @@ static inline int Log2(uint64_t x) {
 
 // Bitmask selecting the k-th bit in a byte
 // static constexpr uint8_t kBitmask[] = {1, 2, 4, 8, 16, 32, 64, 128};
-static constexpr uint8_t GetBitMask(uint8_t index) {
+template <typename T>
+static constexpr uint8_t GetBitMask(T index) {
   // DCHECK(index >= 0 && index <= 7);
+  ARROW_COMPILER_ASSUME(index >= 0 && index <= 7);
   return static_cast<uint8_t>(1) << index;
 }
 
 // the bitwise complement version of kBitmask
 // static constexpr uint8_t kFlippedBitmask[] = {254, 253, 251, 247, 239, 223, 191, 127};
-static constexpr uint8_t GetFlippedBitMask(uint8_t index) {
+template <typename T>
+static constexpr uint8_t GetFlippedBitMask(T index) {
   // DCHECK(index >= 0 && index <= 7);
+  ARROW_COMPILER_ASSUME(index >= 0 && index <= 7);
```

Comment:
Putting the assumption here doesn't have the same effect. Because `signed_index % 8` happens before the call to `GetBitMask` and `GetFlippedBitMask`.

Reply:
I'm not fully understand this, `signed_index % 8` happens before the call to GetBitmask doesn't means the code of `GetBitMask` cannot benifit from this? Or you means the system might analyze `% 8` and do the assume itself?

Reply:
Or from the godbolt link, do you mean change `ClearBit` and `SetBit` to ub if `i < 0`?

Reply:
You're right, tried in godbolt, I'll add this

Reply:
`ARROW_COMPILER_ASSUME(index >= 0 && index <= 7)` is superfluous because shifts are already UB in C when the size is bigger than the bit-width of the integer.

Reply:
Aha, I use `ARROW_COMPILER_ASSUME(index >= 0 && index <= 7)` like `DCHECK` here, removed it